core/headerchain.go

```diff
@@ -105,302 +122,212 @@ func NewHeaderChain(chainDb ethdb.Database, config *params.ChainConfig, engine c
-		alreadyKnown := parentKnown && hc.HasHeader(hash, number)
-		if !alreadyKnown {
-			// Irrelevant of the canonical status, write the TD and header to the database.
-			rawdb.WriteTd(batch, hash, number, newTd)
-			hc.tdCache.Add(hash, newTd)
-
-			rawdb.WriteHeader(batch, header)
-			inserted = append(inserted, numberHash{number, hash})
-			hc.headerCache.Add(hash, header)
-			hc.numberCache.Add(hash, number)
-			if firstInserted < 0 {
-				firstInserted = i
-			}
+		// Delete the header and the block
+		rawdb.DeleteCanonicalHash(hc.headerDb, parent.Number64())
+
+		//TODO: Run state processor to rollback state
+
+		// Add to the stack
+		hashStack = append(hashStack, parent)
+		parent = hc.GetHeader(parent.Parent(), parent.Number64()-1)
+
+		if parent == nil {
+			log.Warn("unable to trim blockchain state, one of trimmed blocks not found")
+			return nil
 		}
-		parentKnown = alreadyKnown
-		lastHeader, lastHash, lastNumber = header, hash, number
 	}
 
-	// Skip the slow disk write of all headers if interrupted.
-	if hc.procInterrupt() {
-		log.Debug("Premature abort during headers import")
-		return &headerWriteResult{}, errors.New("aborted")
+	// Run through the hash stack to update canonicalHash and forward state processor
+	for i := len(hashStack); i > 0; i-- {
+		rawdb.WriteCanonicalHash(hc.headerDb, hashStack[i].Hash(), hashStack[i].Number64())
+
+		//TODO: Run the state processor forward
 	}
-	// Commit to disk!
-	if err := batch.Write(); err != nil {
-		log.Crit("Failed to write headers", "error", err)
-	}
-	batch.Reset()
-
-	var (
-		head    = hc.CurrentHeader().Number[types.QuaiNetworkContext].Uint64()
-		localTd = hc.GetTd(hc.currentHeaderHash, head)
-		status  = SideStatTy
-	)
-	// If the total difficulty is higher than our known, add it to the canonical chain
-	// Second clause in the if statement reduces the vulnerability to selfish mining.
-	// Please refer to http://www.cs.cornell.edu/~ie53/publications/btcProcFC.pdf
-	// TODO: need to change the HLCR logic to return something else if the difficulties are equal (replace the first if statement)
-	reorg := hc.HLCR(localTd, newTd)
-	equalTd := newTd[0].Cmp(localTd[0]) == 0 && newTd[1].Cmp(localTd[1]) == 0 && newTd[2].Cmp(localTd[2]) == 0
-	if !reorg && equalTd {
-		if lastNumber < head {
-			reorg = true
-		} else if lastNumber == head {
-			reorg = mrand.Float64() < 0.5
-		}
+
+	return nil
+}
+
+// Reset purges the entire blockchain, restoring it to its genesis state.
+func (hc *HeaderChain) Reset() error {
+	return hc.ResetWithGenesisBlock(hc.genesisHeader)
+}
+
+// ResetWithGenesisBlock purges the entire blockchain, restoring it to the
+// specified genesis state.
+func (hc *HeaderChain) ResetWithGenesisBlock(genesis *types.Header) error {
+
+	hc.headermu.Lock()
+	defer hc.headermu.Unlock()
+
+	//Iterate through my heads and trim each back to genesis
+	for _, head := range hc.heads {
+		hc.trim(hc.genesisHeader, head)
```

Comment:
Why do you trim? Why not just `SetCurrentHeader(...)` to the Genesis header? Trimming can/should probably all be handled in one place by a "garbage collector"